datastore/prometheus.go

```diff
@@ -0,0 +1,331 @@
+			Namespace: "alaz",
+			Name:      "http_status",
+		},
+		[]string{"fromPod", "fromNamespace", "toPod", "toNamespace", "toHost", "toPort", "toService", "status",
+			"fromAcornProject", "fromAcornApp", "fromAcornContainer", "fromAcornAccountId", "fromAcornAppNamespace",
+			"toAcornProject", "toAcornApp", "toAcornContainer", "toAcornAccountId", "toAcornAppNamespace"},
+	)
+	exporter.reg.MustRegister(exporter.statusCounter)
+
+	// Set up caches
+	go exporter.startCache(exporter.podCache, exporter.podEventChan)
+	go exporter.startCache(exporter.svcCache, exporter.svcEventChan)
+	go exporter.handleReqs()
+
+	go func() {
+		http.Handle("/metricz", promhttp.HandlerFor(exporter.reg, promhttp.HandlerOpts{}))
+		go func() {
+			if err := http.ListenAndServe(":8080", nil); err != nil {
+				log.Logger.Error().Err(err).Msg("error while serving metrics")
+			}
+		}()
+		<-exporter.ctx.Done()
+		log.Logger.Info().Msg("Prometheus exporter stopped")
+	}()
+
+	return exporter
+}
+
+func (p *PrometheusExporter) startCache(cache *eventCache, ch chan Event) {
+	for {
+		select {
+		case <-p.ctx.Done():
+			return
+		case obj := <-ch:
+			cache.set(obj.GetUID(), obj)
+		}
+	}
+}
+
+func (p *PrometheusExporter) handleReqs() {
+	for {
+		select {
+		case <-p.ctx.Done():
+			return
+		case req := <-p.reqChanBuffer:
+			p.handleReq(req)
+		}
+	}
+}
+
+func (p *PrometheusExporter) handleReq(req Request) {
+	fromPod, found := p.podCache.get(req.FromUID)
+	if !found {
+		log.Logger.Warn().Str("uid", req.FromUID).Msg("pod not found")
+		return
+	}
+
+	switch req.ToType {
+	case "pod":
+		toPod, found := p.podCache.get(req.ToUID)
+		if !found {
+			return
+		}
+		p.updateMetricsToPod(fromPod.(*PodEvent), toPod.(*PodEvent), req)
+	case "service":
+		toSvc, found := p.svcCache.get(req.ToUID)
+		if !found {
+			return
+		}
+		p.updateMetricsToSvc(fromPod.(*PodEvent), toSvc.(*SvcEvent), req)
+	case "outbound":
+		p.updateMetricsToOutbound(fromPod.(*PodEvent), req)
+	}
+}
+
+func (p *PrometheusExporter) updateMetricsToPod(fromPod, toPod *PodEvent, req Request) {
+	p.latencyHistogram.With(prometheus.Labels{
+		"fromPod":               fromPod.Name,
+		"fromNamespace":         fromPod.Namespace,
+		"toPod":                 toPod.Name,
+		"toNamespace":           toPod.Namespace,
+		"toService":             "",
+		"toHost":                "",
+		"toPort":                strconv.Itoa(int(req.ToPort)),
+		"fromAcornProject":      fromPod.Labels[projectLabel],
+		"fromAcornApp":          fromPod.Labels[appLabel],
+		"fromAcornAppNamespace": fromPod.Labels[appNamespaceLabel],
+		"fromAcornContainer":    fromPod.Labels[containerLabel],
+		"fromAcornAccountId":    fromPod.Labels[accountIdLabel],
+		"toAcornProject":        toPod.Labels[projectLabel],
+		"toAcornApp":            toPod.Labels[appLabel],
+		"toAcornAppNamespace":   toPod.Labels[appNamespaceLabel],
+		"toAcornContainer":      toPod.Labels[containerLabel],
+		"toAcornAccountId":      toPod.Labels[accountIdLabel],
```

Comment:
IIUC, every unique metrics label:value results in a new time series. If that's correct, should we/is it possible to prune this down?

Reply:
Yeah, it is mostly a tradeoff between how much information we want vs. how high we are willing to let our cardinality become.

Based on the last time that I discussed metrics requirements with Vince, the following of these labels are absolutely necessary:

```
fromPod
fromAcornApp
fromAcornAppNamespace
fromAcornContainer
```

I think the following are worthy to be removed:

```
fromNamespace
toNamespace
toService (never seems to be generated anyway)
fromAcornAccountId
toAcornAccountId
```

Imo, the rest of these, while we may not use them right away, are useful to keep around in case we do want to use them. Though I could comment out all the ones that we are not yet using and just uncomment them later when we want them.

Let me know what you think @njhale

Reply:
For more context, I want to keep most of the `to*` labels around so that we can partially generate the mesh view until I have TCP throughput done, in case that takes me a long time.

Reply:
> Though I could comment out all the ones that we are not yet using and just uncomment them later when we want them.

I like this approach.

Also, if we're not exporting these to DataDog, then the associated dollar cost concern becomes less pressing.

Reply:
Oh yeah, these will not go to DD, but they will go to AMP, so there is still a cost concern.

Reply:
Fixed